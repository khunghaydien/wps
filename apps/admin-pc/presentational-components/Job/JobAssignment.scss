.admin-pc-job-job-assignment {
  &__heading {
    margin: 0 0 5px;
    color: $color-text-modest;
    font-size: 1rem;
  }

  &__action-buttons {
    position: relative;
    margin-bottom: 12px;

    button {
      padding: 0 16px;
      margin-right: 12px;
    }
  }

  &__modify-buttons {
    position: absolute;
    top: 0;
    right: 0;

    button {
      padding: 0 16px;
      margin: 0 0 0 12px;
    }
  }

  .react-grid-Canvas .react-grid-Row {
    cursor: pointer;
  }

  .rdg-selected {
    border: none;
  }

  .react-grid-HeaderCell--frozen,
  .rdg-last--frozen {
    box-shadow: none !important;
  }
}
