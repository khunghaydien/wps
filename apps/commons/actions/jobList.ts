import Api from '../api';
import { catchApiError, loadingEnd, loadingStart } from './app';

export const FETCH_JOB_LIST = 'FETCH_JOB_LIST';

export const CLEAR_JOB_LIST = 'CLEAR_JOB_LIST';

/**
 * ジョブを取得
 */
function fetchJobListSuccess(result) {
  return {
    type: FETCH_JOB_LIST,
    payload: result,
  };
}

export const fetchJobListByParent =
  (targetDate, parentItem = null, existingItems = []) =>
  (dispatch) => {
    const req = {
      path: '/time/job/get',
      param: {
        targetDate,
        parentId: parentItem ? parentItem.id : null,
      },
    };

    dispatch(loadingStart());

    return Api.invoke(req)
      .then((result) => {
        const { jobList } = result;
        const newItems = Object.assign([], existingItems);
        const resultItems = Object.assign([], jobList);
        newItems.push(resultItems);
        dispatch(fetchJobListSuccess(newItems));
      })
      .catch((err) => dispatch(catchApiError(err, { isContinuable: true })))
      .then(() => dispatch(loadingEnd()));
  };

export const clearJobList = () => ({
  type: CLEAR_JOB_LIST,
});
