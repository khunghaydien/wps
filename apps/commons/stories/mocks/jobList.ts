export default [
  [
    {
      id: '1',
      code: '#1',
      name: 'TASK #1',
      parentId: null,
      hasChildren: true,
      isDirectCharge: false,
      selectableTimeTrack: true,
      isGroup: false,
    },
    {
      id: '2',
      code: '#2',
      name: 'TASK #2',
      parentId: null,
      hasChildren: false,
      isDirectCharge: false,
      selectableTimeTrack: true,
      isGroup: false,
    },
    {
      id: '3',
      code: '#3',
      name: 'TASK #3',
      parentId: null,
      hasChildren: false,
      isDirectCharge: false,
      selectableTimeTrack: true,
      isGroup: false,
    },
    {
      id: '4',
      code: '#4',
      name: 'TASK #4',
      parentId: null,
      hasChildren: false,
      isDirectCharge: false,
      selectableTimeTrack: true,
      isGroup: false,
    },
  ],
  [
    {
      id: '1-1',
      code: '#1-1',
      name: 'TASK #1-1',
      parentId: '1',
      hasChildren: false,
      isDirectCharge: false,
      selectableTimeTrack: true,
      isGroup: false,
    },
    {
      id: '1-2',
      code: '#1-2',
      name: 'TASK #1-2',
      parentId: '1',
      hasChildren: false,
      isDirectCharge: false,
      selectableTimeTrack: true,
      isGroup: false,
    },
  ],
];
