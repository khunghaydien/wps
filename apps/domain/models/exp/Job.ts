import Api from '../../../commons/api';

export const DEFAULT_LIMIT_NUMBER = 100;

export type Job = {
  id: string;
  code: string;
  hasChildren?: boolean;
  hierarchyParentNameList: string[];
  name: string;
  parentId?: string;
};

export type JobList = Array<Job>;

/**
 * Get job list
 *
 * @param {?string} empId
 * @param {?string} parentId
 * @param {?string} targetDate
 * @param {number} [limitNumber]
 * @returns {Promise<JobList>}
 */
export const getJobList = (
  empId?: string,
  parentId?: string,
  targetDate?: string,
  limitNumber: number = DEFAULT_LIMIT_NUMBER
): Promise<JobList> => {
  return Api.invoke({
    path: '/exp/job/get',
    param: {
      empId,
      parentId,
      targetDate,
      maxCount: limitNumber,
    },
  }).then((response: { jobList: JobList }) => {
    return response.jobList;
  });
};

/**
 * Currently only used in mobile
 *
 * @param {string} companyId
 * @param {?string} empId
 * @param {?string} [parentId]
 * @param {string} targetDate
 * @param {?string} query
 * @param {number} [recordCount]
 * @returns {Promise<JobList>}
 */
export const searchJob = (
  companyId: string,
  empId: string,
  parentId: string,
  targetDate: string,
  query: string,
  recordCount: number = DEFAULT_LIMIT_NUMBER,
  usedIn?: string
): Promise<JobList> => {
  return Api.invoke({
    path: '/job/search',
    param: {
      companyId,
      empId,
      targetDate,
      query,
      recordCount: recordCount + 1,
      usedIn,
    },
  }).then((response: { records: JobList }) => {
    return response.records;
  });
};

/**
 *  Search job by code or name, currently only used in pc
 *
 * @param {?string} empId
 * @param {?string} [name]
 * @param {string} targetDate for report it's accounting date
 * @param {?string} query keyword contained in either code or name
 * @param {?string} usedIn 'REPORT' | 'REQUEST'
 * @returns {Promise<JobList>}
 */
export const searchJobByKeyword = (
  empId: string,
  name: string,
  targetDate: string,
  query?: string,
  usedIn?: string,
  companyId?: string,
  limitNumber: number = DEFAULT_LIMIT_NUMBER
): Promise<JobList> => {
  return Api.invoke({
    path: '/job/search',
    param: {
      empId,
      targetDate,
      query,
      usedIn,
      companyId,
      recordCount: limitNumber + 1,
    },
  }).then((response: { records: JobList }) => {
    return response.records;
  });
};

export const getRecentlyUsed = (
  targetDate: string,
  employeeBaseId: string
): Promise<JobList> => {
  return Api.invoke({
    path: '/job/recently-used/list',
    param: {
      targetDate,
      employeeBaseId,
    },
  }).then((response: { records: JobList }) => {
    return response.records;
  });
};
