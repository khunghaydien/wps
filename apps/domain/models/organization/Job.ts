export type Job = {
  id: string;
  name: string;
  name_L0: string;
  name_L1: string;
  name_L2: string;
  code: string;
  companyId: string;
  /* invisible in April'21  */
  // jobTypeId?: string;
  // jobType?: {
  //   name: string;
  // };
  departmentId: string;
  department: {
    name: string;
  };
  parentId: string;
  parent: {
    name: string;
    code: string;
  };
  jobOwnerId: string;
  jobOwner: {
    name: string;
    code: string;
  };
  validDateFrom: string;
  validDateTo: string;
  validDateThrough?: string;
  hasChildren: boolean;
  isEditLocked: boolean;
};

export const defaultValue: Job = {
  id: '',
  name: '',
  name_L0: '',
  name_L1: '',
  name_L2: '',
  code: '',
  companyId: '',
  departmentId: '',
  department: {
    name: '',
  },
  parentId: '',
  parent: {
    name: '',
    code: '',
  },
  jobOwnerId: '',
  jobOwner: {
    name: '',
    code: '',
  },
  validDateFrom: '',
  validDateTo: '',
  hasChildren: false,
  isEditLocked: false,
};

export default defaultValue;
