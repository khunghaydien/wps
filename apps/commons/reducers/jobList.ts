import { CLEAR_JOB_LIST, FETCH_JOB_LIST } from '../actions/jobList';

const initialState = [];

export default function jobListReducer(state = initialState, action) {
  switch (action.type) {
    case FETCH_JOB_LIST:
      return action.payload;
    case CLEAR_JOB_LIST:
      return initialState;

    default:
      return state;
  }
}
