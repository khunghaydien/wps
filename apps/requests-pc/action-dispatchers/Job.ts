import { catchApiError } from '../../commons/actions/app';

import { actions as jobActions } from '../../domain/modules/exp/job/list';
import { AppDispatch } from '../modules/AppThunk';
import { actions as dialogLoadingActions } from '../modules/ui/expenses/dialog/isLoading';
import { actions as jobSelectListActions } from '../modules/ui/expenses/dialog/jobSelect/list';

export const getJobList =
  (parentId: string, targetDate: string) => (dispatch: AppDispatch) => {
    dispatch(dialogLoadingActions.toggle(true));
    return dispatch(jobActions.list(parentId, targetDate, undefined))
      .then((result) => {
        dispatch(jobSelectListActions.set([result.payload]));
      })
      .catch((err) => dispatch(catchApiError(err, { isContinuable: true })))
      .then(() => dispatch(dialogLoadingActions.toggle(false)));
  };

export const getRecentJobs =
  (targetDate: string, empId: string) => (dispatch: AppDispatch) => {
    return dispatch(jobActions.getRecentlyUsed(targetDate, empId))
      .then((result) => {
        dispatch(jobSelectListActions.setRecentResult(result.payload));
        return result.payload;
      })
      .catch((err) => dispatch(catchApiError(err, { isContinuable: true })));
  };

export const searchJobs =
  (
    keyword: string,
    targetDate: string,
    employeeId: string,
    companyId?: string
  ) =>
  (dispatch: AppDispatch) => {
    return dispatch(
      jobActions.getJobSearchResult(
        keyword,
        targetDate,
        employeeId,
        'REQUEST',
        companyId
      )
    )
      .then((result) => {
        dispatch(jobSelectListActions.setSearchResult(result.payload));
        return result.payload;
      })
      .catch((err) => dispatch(catchApiError(err, { isContinuable: true })));
  };

export const getJobSearchResult =
  (keyword: string, targetDate: string, employeeId: string) =>
  (dispatch: AppDispatch) => {
    dispatch(dialogLoadingActions.toggle(true));
    return dispatch(searchJobs(keyword, targetDate, employeeId)).then(() =>
      dispatch(dialogLoadingActions.toggle(false))
    );
  };

export const getNextJobList =
  (item: any, items: any, parentId: string, targetDate: string) =>
  (dispatch: AppDispatch) => {
    dispatch(dialogLoadingActions.toggle(true));
    return dispatch(jobActions.list(parentId, targetDate, undefined))
      .then((result) => {
        items.push(result.payload);
        dispatch(jobSelectListActions.set(items));
      })
      .catch((err) => dispatch(catchApiError(err, { isContinuable: true })))
      .then(() => {
        dispatch(dialogLoadingActions.toggle(false));
      });
  };
