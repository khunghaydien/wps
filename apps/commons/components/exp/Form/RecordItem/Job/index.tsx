import React from 'react';

import { Record } from '../../../../../../domain/models/exp/Record';
import { Report } from '../../../../../../domain/models/exp/Report';

import msg from '../../../../../languages';
import LabelWithHint from '../../../../fields/LabelWithHint';
import QuickSearch, { Option } from '../../QuickSearch';

type Props = {
  expRecord: Record;
  expReport: Report;
  hintMsg?: string;
  isFinanceApproval?: boolean;
  readOnly: boolean;
  recordIdx: number;
  getRecentJobs: () => Promise<Option[]>;
  handleClickJobBtn: (arg0: string) => void;
  searchJobs: (keyword?: string) => Promise<Option[]>;
  updateReport: (arg0: string, arg1: any, arg2: boolean) => void;
};

const ROOT = 'ts-expenses-requests__record_job';

const RecordJob = (props: Props) => {
  const {
    expReport,
    expRecord,
    readOnly,
    handleClickJobBtn,
    hintMsg,
    recordIdx,
    isFinanceApproval,
  } = props;
  const recordDate = expRecord.recordDate;
  const { jobId, jobCode, jobName } = expRecord.items[0];

  let recordJobCode = expReport.jobCode;
  let recordJobName = expReport.jobName;
  if (jobId) {
    recordJobCode = jobCode;
    recordJobName = jobName;
  }

  const onSelectJob = (x: Option) => {
    props.updateReport(`records[${recordIdx}].items[0].jobId`, x.id, true);
    props.updateReport(`records[${recordIdx}].items[0].jobName`, x.name, true);
    props.updateReport(`records[${recordIdx}].items[0].jobCode`, x.code, true);
  };

  const displayValue = recordJobCode
    ? `${recordJobCode} - ${recordJobName}`
    : '';

  return (
    <>
      <div className={`${ROOT} ts-text-field-container`}>
        <LabelWithHint
          text={msg().Exp_Lbl_Job}
          hintMsg={(!readOnly && hintMsg) || ''}
          isRequired
        />
        <QuickSearch
          isSkipRecentlyUsed={isFinanceApproval}
          ROOT={ROOT}
          hideRemoveIcon
          disabled={readOnly}
          placeholder={msg().Com_Lbl_PressEnterToSearch}
          selectedId={jobId || expReport.jobId}
          targetDate={expReport.scheduledDate || expReport.accountingDate || ''}
          displayValue={displayValue}
          onSelect={onSelectJob}
          getRecentlyUsedItems={props.getRecentJobs}
          getSearchResult={props.searchJobs}
          openSearchDialog={() => handleClickJobBtn(recordDate)}
        />
      </div>
    </>
  );
};
export default RecordJob;
