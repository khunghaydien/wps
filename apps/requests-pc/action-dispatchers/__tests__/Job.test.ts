import ApiMock from '../../../../__tests__/mocks/ApiMock';
import * as actions from '../Job';
import jobList, { jobSearchResult } from './mocks/jobList.mock';
import Store from './mocks/Store';

jest.mock('uuid/v4', () => () => 1);

describe('Job', () => {
  afterEach(() => {
    ApiMock.reset();
  });

  let store;

  beforeEach(() => {
    store = Store.create();
  });

  describe('getJobList()', () => {
    it('get next level of job list success', async () => {
      ApiMock.mockReturnValue({
        '/exp/job/get': jobList,
      });

      await store.dispatch(actions.getJobList('parentId1', '2020-02-21'));

      expect(store.getActions()).toMatchSnapshot();
    });
  });

  describe('getJobSearchResult()', () => {
    it('search job success', async () => {
      ApiMock.mockReturnValue({
        '/job/search': jobSearchResult,
      });

      await store.dispatch(
        actions.getJobSearchResult('query string', '2020-02-21', '')
      );

      expect(store.getActions()).toMatchSnapshot();
    });
  });

  describe('getNextJobList()', () => {
    it('get next level of job list success', async () => {
      ApiMock.mockReturnValue({
        '/exp/job/get': jobList,
      });

      await store.dispatch(actions.getNextJobList(null, [], '', ''));

      expect(store.getActions()).toMatchSnapshot();
    });
  });
});
