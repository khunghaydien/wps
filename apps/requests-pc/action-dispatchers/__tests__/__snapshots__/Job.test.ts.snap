// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Job getJobList() get next level of job list success 1`] = `
Array [
  Object {
    "payload": true,
    "type": "MODULES/EXPENSES/DIALOG/IS_LOADING/TOGGLE",
  },
  Object {
    "payload": Array [
      Object {
        "code": "EXPENSE_JOB",
        "hasChildren": false,
        "id": "a0i2v00000SSzkmAAD",
        "name": "Expense Job",
        "parentId": "a0i2v00000SSzknAAD",
      },
    ],
    "type": "MODULES/ENTITIES/EXP/JOB/LIST_SUCCESS",
  },
  Object {
    "payload": Array [
      Array [
        Object {
          "code": "EXPENSE_JOB",
          "hasChildren": false,
          "id": "a0i2v00000SSzkmAAD",
          "name": "Expense Job",
          "parentId": "a0i2v00000SSzknAAD",
        },
      ],
    ],
    "type": "MODULES/REQUESTS/DIALOG/JOB_SELECT/LIST/SET",
  },
  Object {
    "payload": false,
    "type": "MODULES/EXPENSES/DIALOG/IS_LOADING/TOGGLE",
  },
]
`;

exports[`Job getJobSearchResult() search job success 1`] = `
Array [
  Object {
    "payload": true,
    "type": "MODULES/EXPENSES/DIALOG/IS_LOADING/TOGGLE",
  },
  Object {
    "payload": Array [
      Object {
        "code": "GENERAL_JOB",
        "companyId": "a0N2v00000VuxjtEAB",
        "department": Object {
          "code": null,
          "name": null,
        },
        "departmentId": null,
        "hasChildren": true,
        "hierarchyParentNameList": Array [],
        "id": "a0i2v00000SSzknAAD",
        "isDirectCharged": true,
        "isScopedAssignment": false,
        "isSelectableExpense": true,
        "isSelectableTimeTrack": true,
        "jobOwner": Object {
          "code": null,
          "name": null,
        },
        "jobOwnerId": null,
        "jobType": Object {
          "code": "DEF_JOB_TYPE",
          "name": "Default Job type",
        },
        "jobTypeId": "a0h2v00000W4JtpAAF",
        "name": "General Job",
        "name_L0": "General Job",
        "name_L1": null,
        "name_L2": null,
        "parent": Object {
          "code": null,
          "name": null,
        },
        "parentId": null,
        "validDateFrom": "2018-06-19",
        "validDateTo": "2100-01-01",
      },
    ],
    "type": "MODULES/ENTITIES/EXP/JOB/LIST_SUCCESS",
  },
  Object {
    "payload": Array [
      Object {
        "code": "GENERAL_JOB",
        "companyId": "a0N2v00000VuxjtEAB",
        "department": Object {
          "code": null,
          "name": null,
        },
        "departmentId": null,
        "hasChildren": true,
        "hierarchyParentNameList": Array [],
        "id": "a0i2v00000SSzknAAD",
        "isDirectCharged": true,
        "isScopedAssignment": false,
        "isSelectableExpense": true,
        "isSelectableTimeTrack": true,
        "jobOwner": Object {
          "code": null,
          "name": null,
        },
        "jobOwnerId": null,
        "jobType": Object {
          "code": "DEF_JOB_TYPE",
          "name": "Default Job type",
        },
        "jobTypeId": "a0h2v00000W4JtpAAF",
        "name": "General Job",
        "name_L0": "General Job",
        "name_L1": null,
        "name_L2": null,
        "parent": Object {
          "code": null,
          "name": null,
        },
        "parentId": null,
        "validDateFrom": "2018-06-19",
        "validDateTo": "2100-01-01",
      },
    ],
    "type": "MODULES/REQUESTS/DIALOG/JOB_SELECT/LIST/SET_SEARCH",
  },
  Object {
    "payload": false,
    "type": "MODULES/EXPENSES/DIALOG/IS_LOADING/TOGGLE",
  },
]
`;

exports[`Job getNextJobList() get next level of job list success 1`] = `
Array [
  Object {
    "payload": true,
    "type": "MODULES/EXPENSES/DIALOG/IS_LOADING/TOGGLE",
  },
  Object {
    "payload": Array [
      Object {
        "code": "EXPENSE_JOB",
        "hasChildren": false,
        "id": "a0i2v00000SSzkmAAD",
        "name": "Expense Job",
        "parentId": "a0i2v00000SSzknAAD",
      },
    ],
    "type": "MODULES/ENTITIES/EXP/JOB/LIST_SUCCESS",
  },
  Object {
    "payload": Array [
      Array [
        Object {
          "code": "EXPENSE_JOB",
          "hasChildren": false,
          "id": "a0i2v00000SSzkmAAD",
          "name": "Expense Job",
          "parentId": "a0i2v00000SSzknAAD",
        },
      ],
    ],
    "type": "MODULES/REQUESTS/DIALOG/JOB_SELECT/LIST/SET",
  },
  Object {
    "payload": false,
    "type": "MODULES/EXPENSES/DIALOG/IS_LOADING/TOGGLE",
  },
]
`;
